server/src/main/java/org/elasticsearch/indices/store/IndicesStore.java

```diff
@@ -320,34 +325,37 @@ private void allNodesResponded() {
                 return;
             }
 
-            clusterService.getClusterApplierService()
-                .runOnApplierThread("indices_store ([" + shardId + "] active fully on other nodes)", Priority.HIGH, currentState -> {
-                    if (clusterStateVersion != currentState.getVersion()) {
-                        logger.trace(
-                            "not deleting shard {}, the update task state version[{}] is not equal to cluster state before "
-                                + "shard active api call [{}]",
-                            shardId,
-                            currentState.getVersion(),
-                            clusterStateVersion
-                        );
-                        return;
-                    }
-                    try {
-                        indicesService.deleteShardStore("no longer used", shardId, currentState);
-                    } catch (Exception ex) {
-                        logger.debug(() -> format("%s failed to delete unallocated shard, ignoring", shardId), ex);
-                    }
-                }, new ActionListener<>() {
-                    @Override
-                    public void onResponse(Void unused) {}
-
-                    @Override
-                    public void onFailure(Exception e) {
-                        logger.error(() -> format("%s unexpected error during deletion of unallocated shard", shardId), e);
-                    }
-                });
+            deleteShardStoreOnApplierThread(shardId, clusterStateVersion);
         }
+    }
 
+    private void deleteShardStoreOnApplierThread(ShardId shardId, long clusterStateVersion) {
+        clusterService.getClusterApplierService()
```

Comment:
I think it wouldn't be necessary to run this on the applier thread, but I wanted to keep the changes minimal.